### PR TITLE
One to Many

### PR DESCRIPTION
**Pull request by charlyraffellini**
Created `PGOneToMany` and `PGKeyValue`.

Added an image to show the motivation for this OneToMany relationship.

image

### CODE REVIEW

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        PGOneToMany<Integer, Integer> pgOneToMany = new PGOneToMany<>();
+
+        pgOneToMany.put(new PGKeyValue<>(SOME_KEY, Optional.empty()));
+
+        assertThat(pgOneToMany.collect((key,values) -> new DummyOneToMany(key,values)),
+                equalTo(Arrays.asList(new DummyOneToMany(SOME_KEY, new ArrayList<>()))));
+    }
+
+    @Test
+    public void collectOneToManyAfterManyPutForSameKey() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: collectMappedValuesAssosiatedWithGivenKey

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        pgOneToMany.put(new PGKeyValue<>(SOME_KEY, Optional.of(SOME_VALUE)));
+        pgOneToMany.put(new PGKeyValue<>(SOME_KEY, Optional.of(SOME_OTHER_VALUE)));
+
+        assertThat(pgOneToMany.collect((key,values) -> new DummyOneToMany(key,values)),
+                equalTo(Arrays.asList(new DummyOneToMany(SOME_KEY, Arrays.asList(SOME_VALUE, SOME_OTHER_VALUE)))));
+    }
+
+    @Test
+    public void collectMultipleOneToManyAfterPutForDifferentKeys() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: You can combine these this test we the above one and change the name to `collectMappedValuesAssosiatedWithGivenKeys`

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    @Test
+    public void collectNothingWhenNoPut() throws Exception {
+        PGOneToMany<Integer, Integer> pgOneToMany = new PGOneToMany<>();
+
+        assertThat(pgOneToMany.collect((key, values) -> new DummyOneToMany(key,values)),
+                equalTo(new ArrayList<>()));
+    }
+
+    @Test
+    public void collectOneToNothingAfterPutKeyWithNoValue() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: notCollectEmptyValues

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import static org.hamcrest.core.IsEqual.equalTo;
+public class PGOneToManyShould {
+
+    private static final int SOME_KEY = 1;
+    private static final Integer SOME_OTHER_KEY = 2;
+    public static final int SOME_VALUE = 3;
+    private static final Integer SOME_OTHER_VALUE = 4;
+
+    @Test
+    public void collectNothingWhenNoPut() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: returnEmptyMappedValuesWhenThereWereNoKeyPairRelationshipDefined

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
+package com.codurance.lightaccess;
+
+import org.junit.Test;
+import java.util.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+public class PGOneToManyShould {
+
+    private static final int SOME_KEY = 1;
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: I would use some string instead of integers. You can give more meaningful values for such a kind of variables.

Comment (robertfirek) [reply src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: Some values are Integers and some of the int. I think want type should be used.

Comment (robertfirek) [reply src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: Using access level modifier should be consistent. All of them should be private

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    private static final String SOME_OTHER_VALUE = "Value 2";
+    private PGOneToMany<String, String> pgOneToMany;
+
+    @Before
+    public void setup() {
+        pgOneToMany = new PGOneToMany<>();
+    }
+
+    @Test
+    public void collectMappedValuesGroupedByKeyWhenKeyIsLinkedToMultipleValues() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: Prosposed name: `collectMappedValuesGroupedByKey`

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        pgOneToMany.put(new PGKeyValue<>(KEY_WITH_SINGLE_VALUE, of(SOME_VALUE)));
+
+        assertThat(pgOneToMany.collect(MappedKeyValues::new),
+                containsInAnyOrder(
+                        new MappedKeyValues(KEY_WITH_MULTIPLE_VALUES, asList(SOME_VALUE, SOME_OTHER_VALUE)),
+                        new MappedKeyValues(KEY_WITH_SINGLE_VALUE, singletonList(SOME_VALUE))));
+    }
+
+    @Test
+    public void collectNoGroupedValuesWhenKeyIsNotLinkedToAnyValue() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: Prosposed name: `collectEmptyMappedValuesWhenAllKeyValuePairsForGivenKeyHaveEmptyValues`

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        pgOneToMany.put(new PGKeyValue<>(KEY_WITHOUT_VALUES, empty()));
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: It would be good to show that the number of empty KeyValue pairs put in the class returns an empty list.

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+        assertThat(pgOneToMany.collect(MappedKeyValues::new),
+                equalTo(singletonList(new MappedKeyValues(KEY_WITHOUT_VALUES, emptyList()))));
+    }
+
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java:47]: I can see another edge case when: You put one or more KeyValue pair with non-empty value and one or more KeyValue pair with an empty value.

Comment (lpalma) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: dfas

**Comment by lpalma** [comment]
#2

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    @Test
+    public void collectEmptyMappedValuesGroupedByKeyWhenAllKeyValuePairsForGivenKeyHaveEmptyValues() throws Exception {
+        pgOneToMany.put(new PGKeyValue<>(KEY_WITHOUT_VALUES, empty()));
+
+        assertThat(pgOneToMany.collect(MappedKeyValues::new),
+                equalTo(singletonList(new MappedKeyValues(KEY_WITHOUT_VALUES, emptyList()))));
+    }
+
+    @Test
+    public void collectMappedValuesGroupedByKeyWhenKeyValuePairsForGivenKeyHaveEmptyAndNonEmptyValues() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: collectMappedValuesOnlyForKeyValuePairsWithNonEmptyValues

File: src/test/java/com/codurance/lightaccess/PGOneToManyShould.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        pgOneToMany.put(new PGKeyValue<>(KEY_WITH_SINGLE_VALUE, of(SOME_VALUE)));
+
+        assertThat(pgOneToMany.collect(MappedKeyValues::new),
+                containsInAnyOrder(
+                        new MappedKeyValues(KEY_WITH_MULTIPLE_VALUES, asList(SOME_VALUE, SOME_OTHER_VALUE)),
+                        new MappedKeyValues(KEY_WITH_SINGLE_VALUE, singletonList(SOME_VALUE))));
+    }
+
+    @Test
+    public void collectEmptyMappedValuesGroupedByKeyWhenAllKeyValuePairsForGivenKeyHaveEmptyValues() throws Exception {
```
Comment (robertfirek) [inline src/test/java/com/codurance/lightaccess/PGOneToManyShould.java]: We can remove `GroupedByKey`
